server.js

```diff
@@ -17,26 +24,36 @@ app.post("/login/oauth/github/token", (req, res, next) => {
     return res.status(400).send("no code");
   }
 
-  axios
-    .post(
-      "https://github.com/login/oauth/access_token",
-      { code, client_id, client_secret },
-      {
-        headers: {
-          Accept: "application/json",
-        },
+  exchangeCodeForToken(code, client_id, client_secret)
+    .then((ghResponse) => {
+      if (ghResponse.error) {
+        return res.status(401).send(ghResponse.error.message);
       }
-    )
-    .then((axiosResponse) => {
-      if (axiosResponse.error) {
-        return res.status(401).send(axiosResponse.error.message);
+      return ghResponse.data.access_token;
+    })
+    .then((token) => getGitHubUser(token))
+    .then((ghUser) => {
+      const user = findUserByUsername(ghUser.login);
+
+      if (!user) {
+        return res.status(401).send("not registered");
       }
 
-      res.json(axiosResponse.data);
+      const token = createJwt(user.username, jwt_secret_key);
+
+      res.json({ access_token: token });
     })
     .catch(next);
 });
 
+app.use(jwtAuthFilter(jwt_secret_key));
+
+// protected routes
+
+app.get("/api/profile", (req, res) => {
+  res.json(res.locals.user);
+});
```

Comment:
You have more control if you set this middleware explicit for your routes.
Otherwise, it might be difficult to outsource some routes.

```js
const authMiddleware = jwtAuthFilter(jwt_secret_key);

app.get("/api/profile", authMiddleware, (req, res) => {
  res.json(res.locals.user);
});
```